fix: Don't load from secrets dir if nonexistent or inaccessible

anoadragon453 opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.

  To start, try providing a short summary of your changes in the Title above. We follow Conventional Commits syntax, please ensure your title is prefixed with one of:
  - `feat: `
  - `fix: `
  - `docs: `
  - `chore: `

  If a section of the PR template does not apply to this PR, then delete that section.

  PLEASE READ:
  -------------------------
  Mealie is moving to a regular, automatic release schedule. This means that all PRs should be in a
  stable state, ready for release. This includes:

  - Ensuring new tests have been added to cover new features, or to prevent regressions.
  - Work is fully complete and usable

 -->

## What type of PR is this?

<!--
  Delete any of the following that do not apply:
 -->

- bug

## What this PR does / why we need it:

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.

  If there is a UI component to the change, please include before/after images.
-->

Mealie has the ability to load settings from files in `/run/secrets`. https://github.com/mealie-recipes/mealie/issues/3852 states that mealie will fail to start up if the `/run/secrets` directory exists, but mealie doesn't have permissions to access it.

This PR modifies the logic to first check if `/run/secrets` exists and that mealie can access it, before handing it off to Pydantic to try and load files from it.

## Which issue(s) this PR fixes:

<!--
If this PR fixes one of more issues, list them here.
One per line, like so:
Fixes #123
Fixes #39
-->

Fixes https://github.com/mealie-recipes/mealie/issues/3852

## Special notes for your reviewer:

<!--
   Is there any particular feedback you would / wouldn't like?
   Which parts of the code should reviewers focus on?
-->

I find the `settings_config_kwargs` dict with a single `SettingsConfigDict` invocation to be cleaner than returning multiple variants of `SettingsConfigDict` in each `if` branch. It also means you can specify the common options only once.

## Testing

<!--
  Describe how you tested this change.
-->

I have tested this in my NixOS config, and verified that it resolves https://github.com/mealie-recipes/mealie/issues/3852.

anoadragon453 committed:
Don't load from secrets dir if nonexistent or inaccessible

anoadragon453 commented on mealie/core/settings/settings.py:
```diff
@@ -34,6 +35,27 @@
+
+    # Likewise, check we have permission to read from the secrets directory.
+    if not os.access(secrets_dir, os.R_OK):
+        print(f"Secrets directory '{secrets_dir}' cannot be read from. Check permissions")
```
This is a false positive.

anoadragon453 commented on mealie/core/settings/settings.py:
```diff
@@ -34,6 +35,27 @@ def determine_secrets(data_dir: Path, production: bool) -> str:
+
+    # Check that the secrets directory exists.
+    if not os.path.exists(secrets_dir):
+        print(f"Secrets directory '{secrets_dir}' does not exist")
```
I'm struggling to figure out how to get access to the `logging` lib here. I'm unable to pass `self` to `get_secrets_dir` as `AppSettings` is not instantiated yet when this is called.

anoadragon453 replied:
I think we'll either need to accept that `print` is OK here, or modify `get_logging` to have an extra argument to not try and call `get_app_settings` when `__root_logger` is `None`.

https://github.com/mealie-recipes/mealie/blob/37d93d4e4b255b01aab4ea7258058c2fc818426a/mealie/core/root_logger.py#L20-L21

michael-genson replied:
Can we split app settings into two parts? Something like:

```python
class AppLoggingSettings(BaseSettings):
    """Subset of AppSettings to only access logging-related settings"""

    TESTING: bool = False
    PRODUCTION: bool

    LOG_CONFIG_OVERRIDE: Path | None = None
    """path to custom logging configuration file"""
    LOG_LEVEL: str = "info"
    """corresponds to standard Python log levels"""


class AppSettings(AppLoggingSettings):
    theme: Theme = Theme()

    BASE_URL: str = "http://localhost:8080"
    ...  # everything else here
```

Then add a new method `get_logging_settings`, and just copy what we have with `get_app_settings`, etc.

anoadragon453 replied:
Thanks! This helped, as did calling `get_secrets_dir` outside of `AppSettings.__init__` by passing `_secrets_dir` as an argument.

anoadragon453 committed:
Split logging settings into separate base class

Doing so allows logging settings to be configured separately
from the main app setting, meaning the logger can be
set up before the rest of the app settings are derived.

michael-genson reviewed (approved):
LGTM!

michael-genson committed:
Merge branch 'mealie-next' into anoa/ignore_secrets_dir
